protocol/morse/protocol.go

```diff
@@ -113,6 +113,43 @@ func (p *Protocol) BuildRequestContext(
 	}, nil
 }
 
+// BuildHydratorRequestContextForEndpoint builds a new request context for a given service ID and endpoint address.
+// This method is used only in the hydrator to enforce performing QoS checks on a specific pre-selected endpoint.
+func (p *Protocol) BuildHydratorRequestContextForEndpoint(
+	serviceID protocol.ServiceID,
+	preSelectedEndpointAddr protocol.EndpointAddr,
+) (gateway.ProtocolRequestContext, error) {
+	endpoints, err := p.getEndpoints(serviceID)
+	if err != nil {
+		return nil, fmt.Errorf("BuildHydratorRequestContextForEndpoint: error getting endpoints for service %s: %w", serviceID, err)
+	}
+
+	// Create a logger specifically for this request context
+	ctxLogger := p.logger.With(
+		"service_id", string(serviceID),
+		"component", "request_context",
+	)
+
+	preselectedEndpoint, ok := endpoints[preSelectedEndpointAddr]
+	if !ok {
+		return nil, fmt.Errorf("BuildHydratorRequestContextForEndpoint: no pre-selected endpoint found for service %s and endpoint address %s", serviceID, preSelectedEndpointAddr)
+	}
+
+	// Create an endpoint map containing only the selected endpoint to ensure the QoS check is performed on the selected endpoint.
+	preselectedEndpointMap := map[protocol.EndpointAddr]endpoint{
+		preSelectedEndpointAddr: preselectedEndpoint,
+	}
+
+	// Return new request context with fullNode, endpointStore, and logger
+	return &requestContext{
+		logger:                   ctxLogger,
+		fullNode:                 p.fullNode,
+		sanctionedEndpointsStore: p.sanctionedEndpointsStore,
+		endpoints:                preselectedEndpointMap,
```

Comment:
`selectedEndpoint` field on the context should  be set for this case (please see the TODO_REFACTOR I added).